wazuh-agent: init at 4.7.3

V3ntus opened the pull request:
## Description of changes

Add wazuh-agent to the NixOS packages repository.

Wazuh is an open source XDR and SIEM for cybersecurity professionals and hobbyists to monitor security events across their agents.
[wazuh.com](https://wazuh.com/)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

V3ntus committed:
maintainers: add V3ntus

V3ntus committed:
wazuh-agent: init at 4.7.3

sjdwhiting committed:
maintainers: add sjdwhiting

* Update maintainer-list.nix

* Update package.nix

* Fix missing semmicolons

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/wazuh-agent-xdr-edr-on-nixos/23648/4

V3ntus committed:
Merge branch 'master' into wazuh-agent

nephalemsec commented:
@V3ntus You are the hero we all needed.

V3ntus commented:
I didn't get them in this PRs coauthors unfortunately, but @nealfennimore and @sjdwhiting were big help in this too

xzfc commented on nixos/modules/services/security/wazuh/generate-agent-config.nix:
```diff
@@ -0,0 +1,169 @@
+{ cfg, pkgs }:
+pkgs.writeText "ossec.conf" ''
```
This file is pretty big and it would hard to keep it in sync with the upstream.
I suggest to use the upstream file and `substitute` a few values.

V3ntus replied:
Agreed, we could simplify this. I'll explore just using substitution.

xzfc commented on nixos/modules/services/security/wazuh/wazuh.nix:
```diff
@@ -0,0 +1,121 @@
+        assertion = !( cfg.agent.managerIP == "" );
+        message = "services.wazuh.agent.managerIP must be set";
+      }
+      {
+        assertion = cfg.agent.managerPort > 0 && cfg.agent.managerPort <= 65535;
+        message = "services.wazuh.agent.managerPort is set to a port out of range";
+      }
```
Use `types.port` instead.

xzfc commented on nixos/modules/services/security/wazuh/wazuh.nix:
```diff
@@ -0,0 +1,121 @@
+    environment.systemPackages = [ pkg ];
+
+    assertions = [
+      {
+        assertion = !( cfg.agent.managerIP == "" );
+        message = "services.wazuh.agent.managerIP must be set";
+      }
```
Use `nonEmptyStr` instead.

xzfc commented on nixos/modules/services/security/wazuh/wazuh.nix:
```diff
@@ -0,0 +1,121 @@
+      after = [ "network.target" "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      preStart = ''
+        rsync -av --exclude '/etc/client.keys' --exclude '/logs/' ${pkg}/ ${stateDir}/
+        cp ${generatedConfig} ${stateDir}/etc/ossec.conf
+
+        find ${stateDir} -type f -exec chmod 644 {} \;
+        find ${stateDir} -type d -exec chmod 750 {} \;
+        chmod u+x ${stateDir}/bin/*
+        chmod u+x ${stateDir}/active-response/bin/*
+        chown -R ${wazuhUser}:${wazuhGroup} ${stateDir}
+      '';
```
Why this is needed?

V3ntus replied:
In an FHS environment, these are the requirements Wazuh would need to operate in. It's a super hacky way to emulate it. I'm sure there are better alternatives, but I'm not familiar enough with Nix to decide on what the best way to do it is.

pluiedev commented on nixos/modules/services/security/wazuh/wazuh.nix:
```diff
@@ -0,0 +1,121 @@
+        managerIP = mkOption {
+          type = types.str;
+          description = ''
+            The IP address or hostname of the manager. This is a required value.
```
Not necessary — if there's no default given, it's required.
```suggestion
            The IP address or hostname of the manager.
```

pluiedev commented on nixos/modules/services/security/wazuh/wazuh.nix:
```diff
@@ -0,0 +1,121 @@
+        };
+
+        managerPort = mkOption {
+          type = types.int;
```
```suggestion
          type = types.port;
```

pluiedev commented on pkgs/by-name/wa/wazuh/package.nix:
```diff
@@ -0,0 +1,249 @@
+  unpackPhase = with lib.strings; ''
+    mkdir -p $out/src/src/external
+    cp --no-preserve=all -rf $src/* $out/src
+    ${(concatStringsSep "\n" (map(dep: "tar -xzf ${dep} -C $out/src/src/external") dependencies))}
```
```suggestion
    ${concatMapStringsSep "\n" (dep: "tar -xzf ${dep} -C $out/src/src/external") dependencies}
```

pluiedev commented on pkgs/by-name/wa/wazuh/package.nix:
```diff
@@ -0,0 +1,249 @@
+    make -C $out/src/src -j $(nproc) -d TARGET=agent INSTALLDIR=$out build
+  '';
+
+  fixupPhase = ":";
```
What's this for...?

V3ntus replied:
Not sure, it was left in from previous work. I'll test without it, if the phase goes through without any bash instructions like this.

pluiedev replied:
> Not sure, it was left in from previous work. I'll test without it, if the phase goes through without any bash instructions like this.

My gut feeling would be that these were used to skip over certain phases, but nowadays that should be done by setting `doCheck`/`doBuild`/... to false.

V3ntus replied:
That makes sense! I'll look into those options then, thank you

V3ntus committed:
ref: use `types.port` for cfg manager port definition

Co-authored-by: Leah Amelia Chen <github@acc.pluie.me>

V3ntus committed:
doc: redundant "required" in manager IP option docstring

Co-authored-by: Leah Amelia Chen <github@acc.pluie.me>

V3ntus commented on pkgs/by-name/wa/wazuh/package.nix:
```diff
@@ -0,0 +1,249 @@
+  ];
+in
+stdenv.mkDerivation rec {
+  pname = "wazuh-agent-build";
```
The package in this PR only builds Wazuh agent so it should probably remain `wazuh-agent-build` although it's possible this could be a boilerplate for Wazuh manager in the future.

V3ntus commented on pkgs/by-name/wa/wazuh/package.nix:
```diff
@@ -0,0 +1,249 @@
+      {
+        name = "libpcre2";
+        sha256 = "sha256-0Lr8NXn6CvCjmVFYbt+jSeH0voPSi+2Gq+Cj/Es0/Po=";
+      })
+    (fetcher
+      {
+        name = "libplist";
+        sha256 = "sha256-iCeNS9/BvWo6GlWk89kzaD0nMroJz3p0n+jsjuxAbjw=";
+      })
+    (fetcher
+      {
+        name = "pacman";
+        sha256 = "sha256-9n3Tiir7NA19YDUNSbdamDp8TgGtdgIFaSDv6EnVsUM=";
+      })
+    (fetcher
+      {
+        name = "libarchive";
+        sha256 = "sha256-+GPzgurZ9hq8Vg/w6tC+OqnpW2+MYnVuHwNPTCOGunk=";
+      })
+    (fetcher
+      {
+        name = "popt";
+        sha256 = "sha256-1ogKBmIsoy3EqjmtXc977y+qgb2TGvvmS6Q0rY/uHao=";
+      })
+    (fetcher
+      {
+        name = "rpm";
+        sha256 = "sha256-TLxfyjwT3y3RUBLGi81SMH0omhvn6TxkjZ4P44ezFnA=";
+      })
+    (fetcher
+      {
+        name = "cpython";
+        sha256 = "sha256-Ct2yYX/g3PnnVvt9gOb1Uffqb/DCLIfcP8rTBva6yrA=";
+      })
+    (fetcher
+      {
+        name = "jemalloc";
+        sha256 = "sha256-KyLoWzUsffVQukCKQiUeUejf+myRqi4ftIBKsxf/vKA=";
+      })
+  ];
```
This seems to be needed for per-Wazuh version dependency resolution https://github.com/wazuh/wazuh/blob/9709e57d92d611612c3f7eb178caf8638d73bf2e/src/Makefile#L673
https://github.com/wazuh/wazuh/blob/9709e57d92d611612c3f7eb178caf8638d73bf2e/src/Makefile#L1345-L1346

V3ntus committed:
Merge branch 'NixOS:master' into wazuh-agent

V3ntus committed:
ref: remove static uid/gid definition for wazuh user/group

V3ntus committed:
ref: remove `with lib` usages
fmt: indent where needed
ref: clean up buildInputs
ref: pname simplified to just wazuh-agent
ref: implemented build working directory
ref: remove use of deprecated, bare `--replace`
ref: migrate to implicit build

TODO: Implicit build introduced odd EOF errors. Setting a `buildPhase = ":"` seems to get rid of this issue, not sure how to proceed.

V3ntus commented on pkgs/by-name/wa/wazuh/package.nix:
```diff
@@ -0,0 +1,249 @@
+    echo "Running make TARGET=agent..."
+    export OSSEC_LIBS="-lzstd"
+    make -C $out/src/src -j $(nproc) -d TARGET=agent INSTALLDIR=$out build
+  '';
```
I have implicit build working in 75f109a71688e6edb77f8e1f97afd5477e12b3fb but this introduces a vague "unexpected end-of-file" error after the derivation completes all phases:
![image](https://github.com/NixOS/nixpkgs/assets/29584664/87f51a5c-a5ad-4513-948e-a73fb73474ec)

V3ntus committed:
fmt: run editorconfig-checker and nixpkgs-fmt

V3ntus committed:
Merge branch 'master' into wazuh-agent

nephalemsec commented:
Whats holding this one back from being merged out of curiosity?

kieeps commented:
> Whats holding this one back from being merged out of curiosity?

Not sure if it's related, but I saw the author mention something about  waiting for wazuh 4.9.0 to be released [here](https://github.com/NixOS/nixpkgs/issues/230623)

V3ntus commented:
Closing as stale. The community is working on init at 4.9.1 at the moment. See #230623
